File: app/src/main/java/com/example/gmailish/data/dao/MailLabelDao.java
package com.example.gmailish.data.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.example.gmailish.data.entity.MailEntity;
import com.example.gmailish.data.entity.MailLabelCrossRef;

import java.util.List;

@Dao
public interface MailLabelDao {

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long add(MailLabelCrossRef ref);

    @Query("DELETE FROM mail_label_cross_ref WHERE mailId = :mailId AND labelId = :labelId")
    int remove(String mailId, String labelId);

    @Query("DELETE FROM mail_label_cross_ref WHERE mailId = :mailId")
    int clearForMail(String mailId);

    // NEW: clear all cross-refs for a given label (used by LabelRepository.deleteLabel)
    @Query("DELETE FROM mail_label_cross_ref WHERE labelId = :labelId")
    int clearForLabel(String labelId);

    // For listing mails by label
    @Query("SELECT m.* FROM mails m INNER JOIN mail_label_cross_ref x ON m.id = x.mailId WHERE x.labelId = :labelId ORDER BY m.timestamp DESC")
    List<MailEntity> getMailsForLabelSync(String labelId);

    // NEW: get labels for a mail (used by offline mail detail building)
    @Query("SELECT labelId FROM mail_label_cross_ref WHERE mailId = :mailId")
    List<String> getLabelsForMailSync(String mailId);
    @Query(
            "SELECT m.* FROM mails m " +
                    "INNER JOIN mail_label_cross_ref x ON m.id = x.mailId " +
                    "WHERE x.labelId = :labelId AND m.ownerId = :ownerId " +
                    "ORDER BY m.timestamp DESC"
    )
    List<MailEntity> getMailsForLabelSync(String labelId, String ownerId);
}

File: app/src/main/java/com/example/gmailish/data/dao/MailLabelDao.kt
package com.example.gmailish.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import com.example.gmailish.data.entity.MailLabelCrossRef
import com.example.gmailish.data.entity.relations.LabelWithMails
import com.example.gmailish.data.entity.relations.MailWithLabels
import kotlinx.coroutines.flow.Flow

@Dao
interface MailLabelDao {
    // Add a mail-label relation (ignore if already exists)
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun add(ref: MailLabelCrossRef)

    // Remove a specific mail-label relation
    @Query("DELETE FROM mail_label_cross_ref WHERE mailId = :mailId AND labelId = :labelId")
    suspend fun remove(mailId: String, labelId: String)

    // Clear all relations for a mail (used when server returns authoritative labels)
    @Query("DELETE FROM mail_label_cross_ref WHERE mailId = :mailId")
    suspend fun clearForMail(mailId: String)

    // Clear all relations for a label (when deleting a label)
    @Query("DELETE FROM mail_label_cross_ref WHERE labelId = :labelId")
    suspend fun clearForLabel(labelId: String)

    // Observe mails for a specific label (reactive)
    @Transaction
    @Query("""
    SELECT * FROM mails
    WHERE id IN (
        SELECT mailId FROM mail_label_cross_ref WHERE labelId = :labelId
    )
    ORDER BY timestamp DESC
""")
    fun observeMailsForLabel(labelId: String): Flow<List<MailWithLabels>>

    // Observe labels for a specific mail (reactive)
    @Transaction
    @Query("""
    SELECT * FROM labels
    WHERE id IN (
        SELECT labelId FROM mail_label_cross_ref WHERE mailId = :mailId
    )
    ORDER BY name COLLATE NOCASE ASC
""")
    fun observeLabelsForMail(mailId: String): Flow<List<LabelWithMails>>
}